addon_imps/tests/storage/test_gitlab.py

```diff
@@ -0,0 +1,211 @@
+
+        with self.assertRaises(ItemNotFound):
+            await self.imp.list_child_items(item_id="1:missing_folder")
+
+    async def test_list_root_items_pagination(self):
+        mock_response_page_1 = [
+            {"id": "1", "name": "repo1", "path_with_namespace": "repo1"},
+            {"id": "2", "name": "repo2", "path_with_namespace": "repo2"},
+        ]
+        self._patch_get(
+            mock_response_page_1,
+            headers={"Link": '<https://gitlab.com/api/v4/projects?page=2>; rel="next"'},
+        )
+        result_page_1 = await self.imp.list_root_items()
+        expected_items_page_1 = [
+            ItemResult(item_id="repo1:", item_name="repo1", item_type=ItemType.FOLDER),
+            ItemResult(item_id="repo2:", item_name="repo2", item_type=ItemType.FOLDER),
+        ]
+
+        self.assertEqual(result_page_1.items, expected_items_page_1)
+        self.assertEqual(result_page_1.next_sample_cursor, "page=2")
+        mock_response_page_2 = [
+            {"id": "3", "name": "repo3", "path_with_namespace": "repo3"},
+        ]
+        self._patch_get(mock_response_page_2, headers={})
+        result_page_2 = await self.imp.list_root_items(page_cursor="page=2")
+        expected_items_page_2 = [
+            ItemResult(item_id="repo3:", item_name="repo3", item_type=ItemType.FOLDER),
+        ]
+
+        self.assertEqual(result_page_2.items, expected_items_page_2)
+        self.assertIsNone(result_page_2.next_sample_cursor)
+
+    async def test_list_child_items_pagination(self):
+        folder_mock_page_1 = [
+            {"name": "src", "path": "src", "type": "tree"},
+            {"name": "README.md", "path": "README.md", "type": "blob"},
+        ]
+        self._patch_get(
+            folder_mock_page_1,
+            headers={
+                "Link": '<https://gitlab.com/api/v4/projects/1/repository/tree?page=2>; rel="next"'
+            },
+        )
+        result_page_1 = await self.imp.list_child_items("1:")
+        expected_items_page_1 = [
+            ItemResult(item_id="1:src", item_name="src", item_type=ItemType.FOLDER),
+            ItemResult(
+                item_id="1:README.md", item_name="README.md", item_type=ItemType.FILE
+            ),
+        ]
+
+        self.assertEqual(result_page_1.items, expected_items_page_1)
+        self.assertEqual(result_page_1.next_sample_cursor, "page=2")
+        folder_mock_page_2 = [
+            {"name": "LICENSE", "path": "LICENSE", "type": "blob"},
+        ]
+        self._patch_get(folder_mock_page_2, headers={})
+        result_page_2 = await self.imp.list_child_items("1:", page_cursor="page=2")
+        expected_items_page_2 = [
+            ItemResult(
+                item_id="1:LICENSE", item_name="LICENSE", item_type=ItemType.FILE
+            ),
+        ]
+
+        self.assertEqual(result_page_2.items, expected_items_page_2)
+        self.assertIsNone(result_page_2.next_sample_cursor)
```

Comment:
Please split this into separate tests to be sure that we have clean environment for each case